lib/Rename/Model/RenameEdit.php

```diff
@@ -0,0 +1,18 @@
+<?php
+
+namespace Phpactor\Rename\Model;
+
+use ArrayIterator;
+
+/**
+ * @extends ArrayIterator<int, LocatedTextEditsMap|RenameResult>
+ */
+class RenameEdit extends ArrayIterator
+{
+    public function __construct(LocatedTextEditsMap|RenameResult|null ...$edits)
```

Comment:
why is this a variadic?

Reply:
Only for enforcing types without using docblock.

Reply:
Ok, let's use the docblock then :)

Reply:
No problem, regardless, I'd like to know why you think this is probably bad practice (especially that in practice the constructor gets max two arguments).

Reply:
because you need to iterate over them and check `instanceof` to get either the rename result or the text edit map. it is implementing the wrong data structure in order to avoid using a docblock.

Reply:
- The order of _documentCnages_ matters thus I used array here (imagine [TextDocumentEdit](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocumentEdit) on `file1` **and then** [RenameFile](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#renameFile) from `file1` to `file2`). Assigning LocatedTextEdits and RenameResult to properties loses the (explicit) order. 
- I removed `null` when converted variadics to docblock. It was used only in tests.

Reply:
Where is this used? and why array_filter here?

https://github.com/przepompownia/phpactor/blob/1dc9c4823f550e26b843a1ad8448fb59714cb5d0/lib/Rename/Adapter/ClassMover/FileRenamer.php#L69-L72

Reply:
> Where is this used?

`FileRenameHandler`

If `WorkspaceRenameEdits` constructor allowed null, `array_filter` were redundant also in the rest calls. `WorkspaceEdit`'s `documentChanges` doesn't allow `null` inside array so here too.

Reply:
I'm still confused. The result of the Promise in `lib/Rename/Adapter/ClassMover/FileRenamer.php` is:

```
            return new WorkspaceRenameEdits(array_filter([
                LocatedTextEditsMap::fromLocatedEdits($locatedEdits),
                new RenameResult($from, $to),
            ]));
```

i.e. why not:

```
            return new WorkspaceRenameEdits(
                LocatedTextEditsMap::fromLocatedEdits($locatedEdits),
                new RenameResult($from, $to),
            );
```

The `array_filter` can be removed, and it doesn't need to be an array?

Reply:
> It doesn't need to be an array?
It's (similarly to WorkspaceEdit's documentChanges elements) the only way to keep the order of operations here.

Reply:
Just to clarify though, it's not needed now?

Reply:
I mean, we are also passing an aggregate of located text edits, rather than the individual edits. So it's a bit ... weird. If this is gojg to be a list of operations then we should or bably not use the LocatedTextEditsMap here, as that' not an atomic operation.

Reply:
I agree that this separation is far from the best and even. I would like to leave some design assumptions to you, but I realize that this may mean a lot of reconstruction (and, in effect, wait possibly long time). Please feel free to replace it with something better. I'm not attached in any way the current form of solution. It would be good to have working such LSP request.